app/services/search/SearchForApplicantService.scala

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2016 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package services.search
+
+import connectors.AuthProviderClient
+import model.Commands.{ Candidate, SearchCandidate }
+import model.Exceptions.{ ApplicationNotFound, ContactDetailsNotFound, PersonalDetailsNotFound }
+import model.PersistedObjects.ContactDetailsWithId
+import org.joda.time.LocalDate
+import repositories._
+import repositories.application.{ GeneralApplicationRepository, PersonalDetailsRepository }
+import uk.gov.hmrc.play.http.HeaderCarrier
+
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext.Implicits.global
+
+object SearchForApplicantService extends SearchForApplicantService {
+  val appRepository = applicationRepository
+  val psRepository = personalDetailsRepository
+  val cdRepository = contactDetailsRepository
+  val authProviderClient = AuthProviderClient
+}
+
+trait SearchForApplicantService {
+
+  val appRepository: GeneralApplicationRepository
+  val psRepository: PersonalDetailsRepository
+  val cdRepository: ContactDetailsRepository
+  val authProviderClient: AuthProviderClient
+
+  def findByCriteria(searchCandidate: SearchCandidate)(implicit hc: HeaderCarrier): Future[List[Candidate]] = searchCandidate match {
+    case SearchCandidate(None, None, None, Some(postCode)) => searchByPostCode(postCode)
+
+    case SearchCandidate(firstOrPreferredName, lastName, dateOfBirth, postCode) =>
+      searchByAllNamesOrDobAndFilterPostCode(firstOrPreferredName, lastName, dateOfBirth, postCode)
+
+    case _ => Future.successful(Nil)
```

Comment:
Do we know how play deals with a failed future? I agree this should be a bad input/error condition because presumably we've received a Search Candidate object that's somehow malformed. I'm not even entirely sure that this default case is reachable